File: packages/sdk/src/modules/params.ts
import {
  ApplyAndTransferParamsStruct,
  ApplyParamsStruct,
  AuditReportStruct,
  InitParamsStruct,
  TransferParamsStruct,
  UpdateParamsStruct,
} from "../artifacts/typechain/src/ConfidentialTransfers.js"
import { ClaimParamsStruct } from "../artifacts/typechain/src/ConfidentialTransfersBridgeable.js"
import { ProofOutput } from "./types.js"
import { Auditors } from "./auditors.js"

export class Params extends Auditors {
  getInitParams(
    output: ProofOutput,
    stateAuditReports?: AuditReportStruct[],
  ): InitParamsStruct {
    return {
      stateAuditReports: stateAuditReports ?? [],
      artifacts: {
        proof: output.proof,
        outputs: output.pubSignals.slice(0, 4),
      },
    }
  }

  getWithdrawParams(
    proofOutput: ProofOutput,
    stateAuditReports?: AuditReportStruct[],
  ): UpdateParamsStruct {
    return this.getUpdateParams(proofOutput, stateAuditReports)
  }

  getDepositParams(
    proofOutput: ProofOutput,
    stateAuditReports?: AuditReportStruct[],
  ): UpdateParamsStruct {
    return this.getUpdateParams(proofOutput, stateAuditReports)
  }

  getApplyParams(
    pendingTransfersIndexes: number[],
    proofOutput: ProofOutput,
    stateAuditReports?: AuditReportStruct[],
  ): ApplyParamsStruct {
    return {
      pendingTransfersIndexes,
      stateAuditReports: stateAuditReports ?? [],
      artifacts: {
        proof: proofOutput.proof,
        outputs: proofOutput.pubSignals.slice(0, 2),
      },
    }
  }

  getTransferParams(
    recipientAddress: string,
    proofOutput: ProofOutput,
    stateAuditReports?: AuditReportStruct[],
    transferAuditReports?: AuditReportStruct[],
    extraData?: string,
  ): TransferParamsStruct {
    return {
      recipient: recipientAddress,
      stateAuditReports: stateAuditReports ?? [],
      transferAuditReports: transferAuditReports ?? [],
      extraData: extraData ?? "0x",
      artifacts: {
        proof: proofOutput.proof,
        outputs: proofOutput.pubSignals.slice(0, 4),
      },
    }
  }

  getApplyAndTransferParams(
    recipientAddress: string,
    pendingTransfersIndexes: number[],
    proofOutput: ProofOutput,
    stateAuditReports?: AuditReportStruct[],
    transferAuditReports?: AuditReportStruct[],
    extraData?: string,
  ): ApplyAndTransferParamsStruct {
    return {
      recipient: recipientAddress,
      pendingTransfersIndexes,
      stateAuditReports: stateAuditReports ?? [],
      transferAuditReports: transferAuditReports ?? [],
      extraData: extraData ?? "0x",
      artifacts: {
        proof: proofOutput.proof,
        outputs: proofOutput.pubSignals.slice(0, 4),
      },
    }
  }

  getClaimParams(
    indexToClaim: number,
    proofOutput: ProofOutput,
    stateAuditReports?: AuditReportStruct[],
  ): ClaimParamsStruct {
    return {
      indexToClaim,
      artifacts: {
        proof: proofOutput.proof,
        outputs: proofOutput.pubSignals.slice(0, 2),
      },
      stateAuditReports: stateAuditReports ?? [],
    }
  }

  private getUpdateParams(
    output: ProofOutput,
    stateAuditReports?: AuditReportStruct[],
  ): UpdateParamsStruct {
    return {
      amount: output.pubSignals[5],
      stateAuditReports: stateAuditReports ?? [],
      artifacts: {
        proof: output.proof,
        outputs: output.pubSignals.slice(0, 2),
      },
    }
  }
}

File: packages/go-sdk/sdk/params.go
package sdk

import (
	"math/big"

	"github.com/ethereum/go-ethereum/common"
	confidential_transfers "github.com/noxlabs/confidential-transfers-go-sdk/sdk/artifacts/contracts/ConfidentialTransfers"
	confidential_transfers_bridgeable "github.com/noxlabs/confidential-transfers-go-sdk/sdk/artifacts/contracts/ConfidentialTransfersBridgeable"
)

type Params struct{}

func (p *Params) GetInitParams(output ProofOutput, stateAuditReports []confidential_transfers.AuditReport) confidential_transfers.InitParams {
	if stateAuditReports == nil {
		stateAuditReports = []confidential_transfers.AuditReport{}
	}
	return confidential_transfers.InitParams{
		StateAuditReports: stateAuditReports,
		Artifacts: confidential_transfers.ZKArtifacts{
			Proof:   output.Proof,
			Outputs: output.PubSignals[:4],
		},
	}
}

func (p *Params) GetMintParams(output ProofOutput, stateAuditReports []confidential_transfers.AuditReport) confidential_transfers.UpdateParams {
	return p.getUpdateParams(output, stateAuditReports)
}

func (p *Params) GetBurnParams(output ProofOutput, stateAuditReports []confidential_transfers.AuditReport) confidential_transfers.UpdateParams {
	return p.getUpdateParams(output, stateAuditReports)
}

func (p *Params) GetWithdrawParams(output ProofOutput, stateAuditReports []confidential_transfers.AuditReport) confidential_transfers.UpdateParams {
	return p.getUpdateParams(output, stateAuditReports)
}

func (p *Params) GetDepositParams(output ProofOutput, stateAuditReports []confidential_transfers.AuditReport) confidential_transfers.UpdateParams {
	return p.getUpdateParams(output, stateAuditReports)
}

func (p *Params) GetTransferParams(recipient common.Address, output ProofOutput, stateAuditReports []confidential_transfers.AuditReport, transferAuditReports []confidential_transfers.AuditReport, extraData []byte) confidential_transfers.TransferParams {
	if stateAuditReports == nil {
		stateAuditReports = []confidential_transfers.AuditReport{}
	}
	if transferAuditReports == nil {
		transferAuditReports = []confidential_transfers.AuditReport{}
	}
	if extraData == nil {
		extraData = []byte{}
	}

	return confidential_transfers.TransferParams{
		Recipient:            recipient,
		StateAuditReports:    stateAuditReports,
		TransferAuditReports: transferAuditReports,
		ExtraData:            extraData,
		Artifacts: confidential_transfers.ZKArtifacts{
			Proof:   output.Proof,
			Outputs: output.PubSignals[:4],
		},
	}
}

func (p *Params) GetApplyParams(pendingTransfersIndexes []*big.Int, output ProofOutput, stateAuditReports []confidential_transfers.AuditReport) confidential_transfers.ApplyParams {
	if stateAuditReports == nil {
		stateAuditReports = []confidential_transfers.AuditReport{}
	}

	return confidential_transfers.ApplyParams{
		PendingTransfersIndexes: pendingTransfersIndexes,
		StateAuditReports:       stateAuditReports,
		Artifacts: confidential_transfers.ZKArtifacts{
			Proof:   output.Proof,
			Outputs: output.PubSignals[:2],
		},
	}
}

func (p *Params) GetApplyAndTransferParams(recipient common.Address, pendingTransfersIndexes []*big.Int, output ProofOutput, stateAuditReports []confidential_transfers.AuditReport, transferAuditReports []confidential_transfers.AuditReport, extraData []byte) confidential_transfers.ApplyAndTransferParams {
	if stateAuditReports == nil {
		stateAuditReports = []confidential_transfers.AuditReport{}
	}
	if transferAuditReports == nil {
		transferAuditReports = []confidential_transfers.AuditReport{}
	}
	if extraData == nil {
		extraData = []byte{}
	}

	return confidential_transfers.ApplyAndTransferParams{
		Recipient:               recipient,
		PendingTransfersIndexes: pendingTransfersIndexes,
		StateAuditReports:       stateAuditReports,
		TransferAuditReports:    transferAuditReports,
		ExtraData:               extraData,
		Artifacts: confidential_transfers.ZKArtifacts{
			Proof:   output.Proof,
			Outputs: output.PubSignals[:4],
		},
	}
}

func (p *Params) GetClaimParams(output ProofOutput, stateAuditReports []confidential_transfers_bridgeable.AuditReport) confidential_transfers_bridgeable.ClaimParams {
	if stateAuditReports == nil {
		stateAuditReports = []confidential_transfers_bridgeable.AuditReport{}
	}
	return confidential_transfers_bridgeable.ClaimParams{
		StateAuditReports: stateAuditReports,
		Artifacts: confidential_transfers_bridgeable.ZKArtifacts{
			Proof:   output.Proof,
			Outputs: output.PubSignals[:2],
		},
	}
}

func (p *Params) getUpdateParams(output ProofOutput, stateAuditReports []confidential_transfers.AuditReport) confidential_transfers.UpdateParams {
	if stateAuditReports == nil {
		stateAuditReports = []confidential_transfers.AuditReport{}
	}

	return confidential_transfers.UpdateParams{
		Amount:            output.PubSignals[5],
		StateAuditReports: stateAuditReports,
		Artifacts: confidential_transfers.ZKArtifacts{
			Proof:   output.Proof,
			Outputs: output.PubSignals[:2],
		},
	}
}
